util/log: signal and reset a message's errorCh on drop

andreimatei opened the pull request:
This patch fixes a deadlock in the bufferSink. Before, if a log message
sent with the forceSync option was dropped by the bufferSink (because of
the maxInFlight limit on the number of queued up flushed of the wrapped
sink), the respective bufferSink.output(msg, opts) call would deadlock
because nobody was signaling the channel on which the result of the
forceSync flush was supposed to come.

This patch fixes it by signaling the respective channel with a specific
error. This error will make it to the caller, telling them that the
forceSync message was not delivered anywhere.

Release note: None

andreimatei committed:
util/log: signal and reset a message's errorCh on drop

This patch fixes a deadlock in the bufferSink. Before, if a log message
sent with the forceSync option was dropped by the bufferSink (because of
the maxInFlight limit on the number of queued up flushed of the wrapped
sink), the respective bufferSink.output(msg, opts) call would deadlock
because nobody was signaling the channel on which the result of the
forceSync flush was supposed to come.

This patch fixes it by signaling the respective channel with a specific
error. This error will make it to the caller, telling them that the
forceSync message was not delivered anywhere.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/82306)
<!-- Reviewable:end -->

andreimatei committed:
util/log: remove TODO about bug that doesn't exist

The bufferSink had a note about a bug, but the bug is not present. The
note was worried about multiple log messages being sent to the sink with
the forceSync flag while a flush is in progress, but that isn't possible
because sink.output(msg, opts) blocks for the duration of the flush for
all forceSync messages and output() is documented to not be called
concurrently.

Closes #72454

Release note: None

knz reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82306#-:-N3Z6WaU3JBm16BzJHdH:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @abarganier and @andreimatei)

___
*[`-- commits` line 24 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82306#-N3Z6HHpDB3bfBmxsg0F):*
nit: forceSink
___
*[`-- commits` line 27 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82306#-N3Z6BFfEfcRSA4BBLHl):*
This "closes ..." need to be in the PR description for it to actually close the issue.
___
*[`pkg/util/log/buffer_sink.go` line 134 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82306#-N3Z6OPr83jNylmYm2iJ:-N3Z6OPr83jNylmYm2iK:bzhz4yg) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba900105b1b0e5786cec18cae5d5252df59a28/pkg/util/log/buffer_sink.go#L134)):*
> ```Go
> 				// since we can't overwrite it in case m.errorCh is also set.
> 				if b.errorCh != nil {
> 					panic("unexpected errorCh already set")
> ```

make that `panic(new.AssertionFailedf(..))`


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82306)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @abarganier and @andreimatei)

___
*[`-- commits` line 24 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82306#-N3Z6HHpDB3bfBmxsg0F):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

nit: forceSink
</blockquote></details>

oh maybe forceSync


<!-- Sent from Reviewable.io -->

andreimatei committed:
util/log: fix a shutdown deadlock in the bufferSink

When the "accumulator" goroutine is done, it sends a signal to the
"flusher" goroutine through a bufferSinkBundle with the `done` flag set.
Before this patch, it was possible for this bundle to be dropped if the
flush buffer was full, resulting in the flusher goroutine never
terminating. This patch fixes it by never dropping a `done` bundle.

Release note: None

andreimatei reviewed (commented):
I've piled on one more commit fixing another similar flush issue.  Please take a look.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82306#-:-N3ZsJt88cDi7oP9cqgQ:b-hxvscs)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @abarganier and @knz)

___
*[`-- commits` line 24 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82306#-N3Z6HHpDB3bfBmxsg0F):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

oh maybe forceSync
</blockquote></details>

how much force would a force sync force if a force sync forced sink sync?
___
*[`-- commits` line 27 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82306#-N3Z6BFfEfcRSA4BBLHl):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This "closes ..." need to be in the PR description for it to actually close the issue.
</blockquote></details>

don't think so, the commit message should do it. We'll see soon enough.
___
*[`pkg/util/log/buffer_sink.go` line 134 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82306#-N3Z6OPr83jNylmYm2iJ:-N3ZmPCQ8PuSzBEjaXlv:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ba900105b1b0e5786cec18cae5d5252df59a28/pkg/util/log/buffer_sink.go#L134)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

make that `panic(new.AssertionFailedf(..))`
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

abarganier reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> - thanks for these improvements! 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82306#-:-N3e5WaMDtP03UCRNJ7Z:bc4p84l)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei and @knz)

___
*[`pkg/util/log/buffer_sink.go` line 68 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/82306#-N3_zVCA9vkoWPrO6D50:-N3_zVCA9vkoWPrO6D51:b7752w9) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9d4a5c091700a38ff909ef2ab9a5ae3e6666508/pkg/util/log/buffer_sink.go#L68)):*
> ```Go
> 	inErrorState bool
> 
> 	// onMsgDrop is a hook that's called, if set, before dropping messages.
> ```

nit: Should we provide some way to set this callback? Or note it as primarily used for tests?



_Code quote:_
```Go
	// onMsgDrop is a hook that's called, if set, before dropping messages.
	onMsgDrop func()
```


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82306)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @abarganier and @knz)

___
*[`pkg/util/log/buffer_sink.go` line 68 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/82306#-N3_zVCA9vkoWPrO6D50:-N3eu0R7DgwkMO20_amF:b-klqij0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9d4a5c091700a38ff909ef2ab9a5ae3e6666508/pkg/util/log/buffer_sink.go#L68)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

nit: Should we provide some way to set this callback? Or note it as primarily used for tests?
</blockquote></details>

meh; this goes away in #82380 anyway


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82306#-:-N3vQcJS7FT5BqaTbXbp:b-ydi9ay)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @abarganier and @knz)



<!-- Sent from Reviewable.io -->
